pdns/slavecommunicator.cc

```diff
@@ -856,10 +869,17 @@ void CommunicatorClass::slaveRefresh(PacketHandler *P)
   time_t now = time(0);
   for(val_t& val :  sdomains) {
     DomainInfo& di(val.di);
+    DomainInfo tempdi;
     // might've come from the packethandler
-    if(!di.backend && !B->getDomainInfo(di.zone, di)) {
+    // Please do not overwrite received DI just to make sure it exists in backend.
+    if(!di.backend) {
+      if (!B->getDomainInfo(di.zone, tempdi)) {
         L<<Logger::Warning<<"Ignore domain "<< di.zone<<" since it has been removed from our backend"<<endl;
         continue;
+      }
+      // Backend for di still doesn't exist and this might cause us to
+      // SEGFAULT on the setFresh command later on
+      di.backend = tempdi.backend;
```

Comment:
Perhaps we should explicitly check that `di.backend` is not `nullptr` after that and bail out otherwise? I don't think `getDomainInfo()` should succeed without setting `DomainInfo::backend` but the check is cheap and would prevent a `nullptr` dereference leading to a crash. And at a quick glance it looks like the `LDAP` backend doesn't set `DomainInfo::backend` in `getDomainInfo()`, but I'll open a separate issue for that.

Reply:
@rgacogne as I recall, the issue is that we have under the nullptr (at 0) a backend class up and running for the default use case. That's why I didn't check the nullptr because it didn't fix the problem when I tried it on a real use case.

Reply:
Right, I'm fine with skipping that check since that should not happen in the first place.